core/src/main/java/net/adoptopenjdk/icedteaweb/resources/downloader/PrioritizedParallelExecutor.java

```diff
@@ -0,0 +1,218 @@
+package net.adoptopenjdk.icedteaweb.resources.downloader;
+
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+/**
+ * Executor which will execute multiple {@link Callable Callables} in parallel.
+ * The callables must be sorted and have the highest priority callable as the first in list.
+ * Executions of lower priority are canceled as soon as a higher priority result has been successfully been returned.
+ * The final {@link Future} will complete as soon as the first callable completed successfully and all other callables
+ * with higher priority (if any) have completed exceptionally
+ */
+public class PrioritizedParallelExecutor {
+
+    private final ExecutorService executor;
+
+    /**
+     * Constructor
+     *
+     * @param executor the executor service which is used for executing the callables.
+     */
+    public PrioritizedParallelExecutor(ExecutorService executor) {
+        this.executor = executor;
+    }
+
+    /**
+     * Executes the passed callables in parallel
+     * The callables must be sorted by priority with the highest priority as the first in list.
+     * Executions of lower priority are canceled as soon as a higher priority result has successfully been calculated.
+     * The returned {@link Future} will complete as soon as the first callable completed successfully and all other
+     * callables with higher priority (if any) have completed exceptionally.
+     *
+     * @param callables the callables to execute.
+     * @return a future holding the best result from all callables.
+     */
+    public <V> Future<V> getSuccessfulResultWithHighestPriority(List<Callable<V>> callables) {
+        if (callables == null || callables.isEmpty()) {
```

Comment:
Maybe you do an assert first and than do the empty check. Otherwise you do not know if it is null or empty based on the exception

Reply:
Can do but there is basically no difference in action to take.
It means one of the caller passed in non-sense and we need to fix it

Reply:
the difference is that the exception will give you an hint if its empty or null. But this is only a minor comment and not mandatory ;)